Practical Flow Ratio Calibration Test

pi-squared-studio opened the pull request:
# Practical Flow Ratio Calibration Test

<img width="600" alt="image" align="center" src="https://github.com/user-attachments/assets/9d64cf7f-2803-4aea-a819-7ff38c92096f" />

## What is this Practical calibration test for?
This minimalistic test is used to quantify and calibrate the actual flow rate of the extruder for the required filament with its typical printing conditions.

<img width="1000" alt="image" src="https://github.com/user-attachments/assets/114aa8c3-084f-4ebd-83c0-c41512402cfa" />

This test uses the current settings of the filament profile, printer, and printing conditions and builds its model based on them. 
The model is a rectangular block with a solid unidirectional filling on a leveling substrate. The filling density is dynamic and depends on the current filament settings. In other words, the 1.0 consumption coefficient in this test will be equal to a solid infill of 100% density and it corresponds to the internal Orca math algorithm for flow rate calculating. The change in flow is determined by the volume of the outgoing extruder material, and it linearly depends on the line width.

Now, knowing the linear rate of flow change along the length of the model and using a conventional ruler, we can determine the required flow rate of the extruder with an accuracy of one thousandth. 

By only one this test, it is possible to really select the required filament consumption not only to ensure the maximum allowable strength of infill, but also to ensure the highest quality surface. Thanks to the flexible configuration of this calibration test, it is possible to perform both fast evaluation tests with minimal consumption of material and printing time and rather complex and accurate ones. 

## Test conditions

<img width="428" height="591" alt="image" src="https://github.com/user-attachments/assets/576e2993-1e23-4fee-9095-2dec493d2ec3" />

> [!CAUTION]
> To perform the test successfully, first calibrate the temperature of the hotend and the cooling conditions of the filament. The printing surface should be as smooth and even as possible (perhaps even shiny). Poor adhesion between layers and lines can lead to objective defects that complicate the readability of the test. 

### Model Width & Depth
You can specify the desired size of the model for the test in order to conveniently evaluate the quality of the result.
The width (along X axis) of the model varies in values: 100, 150 and 200 mm.
The depth (along Y axis) of the model varies in values: 10, 15 and 20 mm.

> [!NOTE]
> Larger sizes entail a higher consumption of filament and the time spent on printing.

### Number of calibration layers
Since this test is with cumulative effect, a larger number of layers will increase the visual effect of internal overflow with excess material. You can specify a minimum value of 4 and a maximum value of 40. The default value is 10.

> [!NOTE]
> Larger number entail a higher consumption of filament and the time spent on printing.

> [!Tip]
> Use a smaller flow range and a larger number of calibration layers to compensate for the loss of readability.

### Start & End flow rate values
These are the numbers of the variable flow range for the test. 

> [!TIP]
> For easier interpret the result later use values of comparable rank to make it. For example, pairs of values like 0.95 and 1.05 (or 0.8 and 1.0) will be easier to recalculate with ruler. The best option would be to enter numbers that are separated from 1.0 with positive and negative delta.

> [!WARNING]
> Please note that the current flow ratio for the extruder/filament will be applied for the actual value of 1.0 for this test. What will remind you the notice label in the dialog box and on the printed ruler. 

### Interlaced

This setting changes the printing mode from progressive to interlaced. However, these methods give slightly different results.

- **Progressive** - The filling is performed from line to line from the low flow value to the high one.  Simulates the conditions closest to the usual ones in 3D printing. The results are slightly smoothed out, as there is a degree of freedom where the plastic can move without causing overflow. The flow detection zone is also well formed to ensure the best geometry of the object.
- **Interlaced** - The printing mode for high-strength products, when the lines are filled through one. 2 passes are formed. The first is the free application of lines, the second is cementing in a limited space. In this mode, the slightest overflow of plastic causes a clearly visible defect.

### Use Z-Hop

Enables the Z-axis lifting mode when changing the line on upper surface. It is used to keep the its clean. For the lowest layer, Z-Hop is switched on constantly to ensure the formation of a more reliable base.

> [!NOTE]
> If Z-Hop mode on, the nozzle lift every time when a new line will draw. But in this case, the formation of retract threads or strings is possible upon surface.
> If Z-Hop mode off, for the upper surface determines the test option when the transfer nozzle to a new line leaves a mark on the excessively extruded material. This helps to visually highlight the slightest defects.

### Print Scale

You can print a scale along with the test on the bed surface, which will help you immediately assess the range of flow rate applied to the test.

<img width="823" height="146" alt="image" src="https://github.com/user-attachments/assets/43a73fbc-75e4-42f6-974b-e429e01f61dd" />

- In the corners there are labels with the start and end values of the flow rate.
- It is also possible to display labels with intermediate flow values.
- The applied scale contains major and minor divisions for easy reading.

### Print Ruler

You can print a ruler along with the test on the bed surface, with which you can easily calculate the actual flow rate value. Each major division shows a distance of 1 cm. Minor divisions show the fractional parts of this dimension.

<img width="823" height="131" alt="image" src="https://github.com/user-attachments/assets/a7d99462-7886-4142-9bdb-8ed426bc7aa9" />

- <img width="359" height="47" alt="image" src="https://github.com/user-attachments/assets/4613cba1-a4d2-41ce-8108-ce125c3d7eb8" /> - it resembles a shortcut for recalculating values. If possible, it will show a convenient fractional formula for recalculation.
- <img width="376" height="54" alt="image" src="https://github.com/user-attachments/assets/253a78a6-1e99-41a6-88d7-5a4389d5e3d5" /> - the label of material type and some printing conditions. Specifies the type of used filament, the printing speed, the method of applying infill (p - progressive, i - interlaced), as well as the current flow coefficient for the selected filament.
- <img width="251" height="65" alt="image" src="https://github.com/user-attachments/assets/d26a700a-d366-410c-a238-34e773c5d5ac" /> - the scale ratio bar.
- The applied scale contains major and minor divisions for easy reading.

> [!NOTE]
> The printable scale and ruler can display a triangular mark <img width="37" height="34" alt="image" src="https://github.com/user-attachments/assets/b6f31fc1-aaff-4bd5-b8ff-989cd316362b" /> that indicates the mechanical extruder/filament flow ratio equal to 1.0. In other words, there would the position of the normal flow ratio of the extruder be if it didn't normalize in any way. Let's say the current filament flow rate is set to 0.96, then this mark would be point to the value 1/0.96 = 1.042. 

## How to make the Practical Flow Ratio Test

1. Open a new project in Orca.
2. Install the needed print, printer, and filament profiles.  Make all necessary changes to the profiles settings. 
> [!WARNING]
> When generating a model, this test will make changes to some settings (for example, disabling Fuzzy Skin) for successful testing. Do not save these settings under already saved profile names, as they may not be desired.
3. Select from the menu "Calibration" > "Flow Ratio" an element "Practical Flow Ratio Test".
4. In the dialog box that opens, set the necessary test settings.
5. Click "OK" and the required model will be generated.
6. Slice the model, then transfer the generated G-Code to the printer for testing.
7. For remake new test, you can change the settings in print, printer, and filament profiles, then run a new test instance without saving the old one.

> [!CAUTION]
> Applying to the conditions of this test will changes a some of the print, printer and filament profiles settings. Do not save these profiles, as this may lead to their loss. After finish the test, memory the measured values, and then set them in needed project after completely resetting or unloading the test.

## How to perceive the results obtained

The visible manifestations may vary depending on the filament used and the printing conditions. 
It can be noted that increasing the number of calibration layers cumulative highlights any visual defects better.

<img width="1864" height="480" alt="image" src="https://github.com/user-attachments/assets/4c0b08ce-fe4d-4e66-bde5-bc14c1fc07d8" />

Visible signs:
1. The area with the best filling. Smooth and even. It does not always have a gloss, it can be matte or gradient. 
2. Loose area with cavities, depressions and uneven lines. There is a clear shortage of material.
3. Plowed and raised surface.  Its level is noticeably higher than normal. Obvious signs of an excess of material.
4. Visible curbs protruding above the surface.
5. Visible excess material collected by the nozzle.
6. Top roller of excess material. This is where the printing line ends, and all the excess printing material accumulates here.
7. Bottom roller of excess material. Surface material accumulates here, which is collected during the return stroke of the nozzle. It is not always a sign of overflow.
8. Deformation of the side wall caused by the pressure of excess material.

The test readings can be divided into 3 boundaries:

- I zone. A clear lack of flow (flow < 1.0).  
> <img width="400" alt="image" src="https://github.com/user-attachments/assets/6700714a-cb7d-478d-b1e5-66be635667ed" />
>
> With this filling, the lines do not affect each other, so it allows you to keep the shape of the printed surface.  This area usually has a smooth, even, and often glossy surface. The surface quality does not suffer from the number of calibration layers, as the material will not be enough to deform in any case. With large values of material shortage, it is possible for it to hole upwards, as the lines lose their lateral support in the form of neighboring lines, or the formation of shrinkage cavities.

- II zone. The flow is normal (flow about 1.0). 
> <img width="400" alt="image" src="https://github.com/user-attachments/assets/acf3f16f-7de0-4d94-bf99-94aeb24063af" />
>
> In this area, the empty pores are filled with additional material. Since the process becomes complex due to the use of excessive pressure, and the formation of lines may not be as stable. The print surface is still strait, but it is no longer as glossy and a little rough. You can determine any number of nominal flow rate in this range for yourself, depending on the desired result -  from comply with the size and quality of the surface to ensure maximum rigidity. Or you can leave the average value.

> [!NOTE]
> The magnitude of this transition zone can vary from a complete absence to a few percent of flow change. If it is impossible to clearly define the boundaries, focus on the "for rigidity" indicator. 

- III zone. Overflow (flow >1.0+)
> <img width="400" alt="image" src="https://github.com/user-attachments/assets/c854a3c8-b71f-4fb3-806c-ee88b25cbf83" />
>
> This is an extremely high amount of plastic that can ruin your prints. Excessive amounts of plastic tend to be squeezed out through the printed walls or come out onto the top surface. The lines turn out to be shapeless, and the excess may collide with the nozzle when printing an adjacent line. This leads to the formation of shapeless flakes of plastic on the surface, the formation of tracks from the nozzle on it, and in the worst case, the model may peel off from the bed.

The boundaries between these zones are the most optimal values for printing:

- Zones I/II - they show the optimal flow ratio for the formation of smooth and even surfaces. Choose this indicator to ensure the best geometry and appearance of the model.
- Zones II/III - determines the best filling of the material between the lines. With this flow rate, conditions are provided for printing the most durable models.

> [!TIP]
> If it is possible to adjust the flow rate separately for the surface and for the internal continuous filling, then you can set the value II/III for the main flow of the extruder/filament, and for the surface the value is lower by this difference.
> For example: II/III =1.05, I/II = 0.96
> The resulting flow value for the surface will be 1 - (1.05 - 0.96) = 0.91

> [!TIP]
> To determine the flow rate for the maximum strength of the solid infill, it is sufficient to perform a test with the number of calibration layers equal to the maximum the number of solid layers for the printed model. Such a test will show the optimal possible infill density that does not lead to critical consequences.

## Assessment methods

- visual: surface quality assessment
- for gloss: it is possible to determine where the smoothest surface is without gaps between tracks and micro-jaggs or ripples on the surface caused by excessively extruded plastic.
- side light: highlights minor visual defects well
- nail or toothpick test: detects the grooves between the lines very well. Swipe your fingernail over the surface, and in the most optimal place you will feel almost no vibrations. The same can be done with the side surface, or with the edge between them.
- a "roller" made of excessive material: since printing takes place in one direction, this excessive material can pile up at the end of the path, which leads to its increase during the test.
- outer wall test: аs mentioned above, too much material can deform the outer perimeter.

## Measurements

Photography often cannot convey that fine line of transition from one measuring zone to another. This line is more clearly visible to the eyes. Use a small, bright source of contrasting light, such as a flashlight, to make these visual differences more visible. It is a good idea to direct a beam of light along the surface of this sample so that the minimum printing defects in height become as readable as possible.

There are 2 main ways to determine the required value:

- use the ruler (physical or printed on the bed) <img width="400" alt="image" src="https://github.com/user-attachments/assets/7dfcf1a4-7ee8-457f-8e80-85134434a268" />

> <img width="800" alt="image" src="https://github.com/user-attachments/assets/f572eb61-eff2-41cb-9f27-4b0ad9c00f6f" />
> Attach a regular ruler (or use the printed one) to measure the distance from the edge of the dough to the desired point.
>
> Let's say we liked 2 points: 2.2cm for position 1|2 (for geometry), and 6.1cm for position 2|3 (for rigidity);<br/>
> knowing the initial flow value of 0.9 and the conversion factor of 1cm = 0.02, we are easily calculating the 2 necessary values:<br/>
>  0.9 + (0.02 * 2.2) = 0.9 + 0.044 = 0.944 (for Geometry)<br/>
>  0.9 + (0.02 * 6.1) = 0.9 + 0.122 = 1.022 (for Rigidity)<br/>
>  0.9 + 0.02 * (2.2 + 6.1) / 2 = 0.9 + 0.083 = 0.983 (Average)<br/>

- use the printed scale on the bed, and directly read the readings from it <img width="400" height="626" alt="image" src="https://github.com/user-attachments/assets/c5f24b1e-33d4-4a67-8404-9627e62a0058" />

pi-squared-studio committed:
GoldenRatio Flow Calibration Test

MxBrnr commented:
Need to verify how easily the results of this test are rendered inaccurate by things such as incorrect z-offset, warping due to being too close to the edge of the build plate (small printers like the Voron 0 may have issues with this due to the size of the heated bed and test piece), etc. One of the biggest reasons the currently implemented test design works is because it removes these variables, eliminating avoidable user-error.

pi-squared-studio commented:
@yw4z 
Yes, it is possible to easily refine this, but first I would like to get a response from the community on how convenient the testing method is. I just have a problem, it seems with the filament, and that this method of flow measuring and the classical one give similar negative results, that there is a 9% overextrusion. If I print with a flow coefficient of 0.91, the model turns out to be very fragile, but at the same time it has amazing geometric parameters. I attribute this fact to wet plastic.

@RF47  
Yes, I took this fact into account. First, I print 2 loose layers - the first with 90% filling, the second with 80%. At least my 3rd layer was already quite smooth. However, it is also possible to change the test model in the future to suit possible nuances.

RF47 commented:
![20251017_193031](https://github.com/user-attachments/assets/9bcebdcc-93bb-4782-883e-86cd43f8cf61)
I can't detect a clear sweet spot (0,96 maybe) ; perhaps 10 mm isn't enough for the flow to stabilize (at least on my machine, with a Bowden system).
is ABS by the way.

pi-squared-studio committed:
Add menu and some usabilities

pi-squared-studio committed:
Merge branch 'main' into Golden-Ratio-Flow-Calibration-Test

pi-squared-studio committed:
Update

pi-squared-studio committed:
Update

pi-squared-studio committed:
Update Plater.hpp

pi-squared-studio commented:
@SoftFever I updated this PR, but for some reason it is not compiled by the system. I have installed the latest version of Orca at my computer, and this PR works well in this project. I would like to further refine this PR, for example, to add a measuring printed scale, but it will be difficult to do this with this problem. This project is completed at this stage, and it can be merged, but as @dewi-ny-je said, it's better to give it a new name. If it is agreed, I will add it to this PR and mark it for draft.

pi-squared-studio committed:
Merge remote-tracking branch 'upstream/main' into Golden-Ratio-Flow-Calibration-Test

SoftFever commented:
I'm still troubleshooting this issue

SoftFever commented:
> I'm still troubleshooting this issue

Okay, this is not the unusual CI/CD issue I'm troubleshooting. 
In this case, there are genuine build errors in the changes that you need to fix.

dewi-ny-je commented:
Actually you already have it, you posted the new test with 98% as best multiplier.
The old pattern with the same filament settings could be useful to see how the -2 pattern looks like.

pi-squared-studio commented:
I don't even know how to check yet. The test is so accurate that it gives almost a 100% result. 
Of course, you can change the transmission extruder coefficient and make a test in non-ideal conditions, but this will be a completely different test.

![IMG_20251031_203047](https://github.com/user-attachments/assets/b415eadb-14e4-40eb-b6f6-ff208dcdb6ed)
For comparison with the above test, a PETG with non-interlaced filling is used. The readings are smoother, but still readable.

![IMG_20251031_222557](https://github.com/user-attachments/assets/0c550d91-c542-4bb1-9686-1c8c8d710597)
And this is a PLA filament

pi-squared-studio committed:
Big Update: +Menu

pi-squared-studio committed:
Update calib_dlg.cpp

pi-squared-studio commented:
@boromyr if I understand correctly, compensation layers exist and there are as many as 3 of them in the new implementation.  As the latest printed samples have shown, their effect is very remarkable. The very first subsequent layer turns out to be very smooth.

In general, the point of this test is to show the filling of the material not on a plane, but in volume. Moreover, the overflow error will accumulate in a large number of layers and become more noticeable in the form of defects on the surface.

dewi-ny-je commented:
Something seems wrong.
<img width="2053" height="1851" alt="image" src="https://github.com/user-attachments/assets/110f9171-f507-4de8-9e94-96d71114c203" />
<img width="832" height="768" alt="image" src="https://github.com/user-attachments/assets/16d27165-4555-4143-9cfd-d489c4ca9740" />

In the middle, where I should have 55 * 0.2 * 0.42 = 4.62 mm3/s (middle so flow multiplier = 1.0), I have instead flow 3.76.

Compare with YOLO which shows correctly about 4.4 mm3/s for "0" because flow is limited by filament properties to 5 mm3/s and current flow multiplier is 0.95

<img width="1649" height="705" alt="image" src="https://github.com/user-attachments/assets/788c654f-8df3-4fd2-8ea1-3ba2a835cb94" />

pi-squared-studio commented:
It is possible that the actual speed is reduced in such a short section, or other unaccounted-for processes are occurring. I don't see any printing errors using this method yet.

pi-squared-studio committed:
Update config

pi-squared-studio commented:
@dewi-ny-je 
Here the math https://manual.slic3r.org/advanced/flow-math

When printing, there are always voids between the coils that cannot be filled with the material in any way. According to calculations, they make up 12% of the total volume, but in practice they may differ both up and down. It all depends on the material and its physical properties. 

If we estimate the actual calculation with a slicer, then at a speed of 60mm/s at h=0.2 w = 0.4, the flow rate will be 60*0,4*0,2 = 4,8mm3/s and in the real ratio -12% => 4.22mm3/s - as shown in the pictures above.

pi-squared-studio commented:
A small announcement

<img width="928" height="521" alt="image" src="https://github.com/user-attachments/assets/e63de4f7-65f9-4190-bd58-7d7270fe84dd" />

boromyr commented:
> <img alt="image" width="928" height="521" src="https://private-user-images.githubusercontent.com/189209038/514758100-e63de4f7-65f9-4190-bd58-7d7270fe84dd.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjMyNDQ0MjAsIm5iZiI6MTc2MzI0NDEyMCwicGF0aCI6Ii8xODkyMDkwMzgvNTE0NzU4MTAwLWU2M2RlNGY3LTY1ZjktNDE5MC1iZDU4LTdkNzI3MGZlODRkZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMTE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTExNVQyMjAyMDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNGZlZTFjMWFkNDdkOWE1YWUwNjMxNzc0YmQ2ZjZmODMyMDMwZWQ2MjAwZWY5YmM2NThkZDRjN2I0MzI0OWY2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.PsEnXhIdHzLNu_9JdlFZmUUcA2y0YOovZkzeTqNJ0uE">
Will it be optional? I don't think it's easy to clean all those letters and those filaments without a vacuum cleaner. Wouldn't it be better to add a single solid layer to get a single piece?

pi-squared-studio commented:
> Will it be optional? I don't think it's easy to clean all those letters and those filaments without a vacuum cleaner. Wouldn't it be better to add a single solid layer to get a single piece?

Yes, this option is disabled. It is used so that you can run quick tests with known parameters.

As for the thin plastic, I also encountered poor adhesion to the bed of fine lines on the first layer. I thought that preparing a solid field would reduce readability. Therefore, I have improved it to such conditions.

<img width="440" height="676" alt="image" src="https://github.com/user-attachments/assets/35c98cca-b765-4418-b1a3-bb767727607f" />

<img width="1079" height="553" alt="image" src="https://github.com/user-attachments/assets/359a2608-7046-4551-ae39-158ccaa15ad2" />

pi-squared-studio committed:
pre-finish

pi-squared-studio committed:
Merge branch 'main' into Golden-Ratio-Flow-Calibration-Test

pi-squared-studio commented:
@SoftFever  @ianalexis  Hi. I've finished this job. Look, if you like everything, then you can take it to the association. 
I can also write instructions, but since I can write many things incorrectly again, it would be better if I put it together on the first page with all the pictures and explanations.

dewi-ny-je commented:
In particular the variation of line width Vs variation in flow multiplier should be checked, since every other flow calibration test offered ao far is changing flow multiplier (the parameter we are trying to tune!) and not the line width, while the gcode preview of this test show a change in line width along the left to right direction

pi-squared-studio committed:
Update config

pi-squared-studio committed:
zHop updates

pi-squared-studio committed:
finally done

pi-squared-studio committed:
some corrections

pi-squared-studio commented:
@SoftFever @ianalexis 
Hurray! I finally finished this PR!
Yes, I even wrote a detailed wiki with all the pictures, explanations and examples on the title page. There may be a misspelling there. If anything, please correct it.

ianalexis commented:
> @SoftFever @ianalexis Hurray! I finally finished this PR! Yes, I even wrote a detailed wiki with all the pictures, explanations and examples on the title page. There may be a misspelling there. If anything, please correct it.

I can help you with the wiki, you will have to create a PR here:
https://github.com/OrcaSlicer/OrcaSlicer_WIKI

pi-squared-studio committed:
fix build / add usabilities

pi-squared-studio committed:
fix

pi-squared-studio committed:
fix 2

pi-squared-studio committed:
fix 3

pi-squared-studio committed:
fix 4

pi-squared-studio committed:
fix 4

pi-squared-studio committed:
Merge branch 'main' into Golden-Ratio-Flow-Calibration-Test

boromyr commented:
The PR has merging issues with the current dev version, is it possible to resolve? I often used this calibration method.

pi-squared-studio committed:
Merge remote-tracking branch 'upstream/main' into Golden-Ratio-Flow-Calibration-Test

pi-squared-studio committed:
Last version update

pi-squared-studio committed:
Merge branch 'main' into Golden-Ratio-Flow-Calibration-Test

pi-squared-studio commented:
@boromyr Yes, I did it. By the way, I use this method all the time, and it's quite accurate. Over the past 5 months, I've conducted dozens of tests with various materials, and it's never let me down. 
It's worth noting that after adjusting the settings for PLA (96%), I achieved such a smooth top surface on the cube that I didn't even need to perform any ironing.

@SoftFever  I have deleted the internal setting  "calib_flowrate_topinfill_special_order". I rewrote the code so that the calibration parameters are passed to the printing module. Therefore, the calibration signs and their settings can be read in full for any test. This will allow for more flexible calibration tests.
The functionality of the classic calibration test is also preserved.

SoftFever commented:
> @boromyr Yes, I did it. By the way, I use this method all the time, and it's quite accurate. Over the past 5 months, I've conducted dozens of tests with various materials, and it's never let me down. 
> 
> It's worth noting that after adjusting the settings for PLA (96%), I achieved such a smooth top surface on the cube that I didn't even need to perform any ironing.
> 
> 
> 
> @SoftFever  I have deleted the internal setting  "calib_flowrate_topinfill_special_order". I rewrote the code so that the calibration parameters are passed to the printing module. Therefore, the calibration signs and their settings can be read in full for any test. This will allow for more flexible calibration tests.
> 
> The functionality of the classic calibration test is also preserved.

Thank you! Will review ASAP

boromyr commented:
@pi-squared-studio Have you noticed that auto slice does not work? Moreover, if you select the interlaced option, it remains active even if you change the model or start a new project.
<img width="1319" height="1030" alt="image" src="https://github.com/user-attachments/assets/50ea5620-7e9d-417b-bdb0-69cf047ae9a7" />

SoftFever committed:
Merge branch 'main' into Golden-Ratio-Flow-Calibration-Test

SoftFever commented:
Hi @pi-squared-studio 

Overall looks good to me.

One small change needed: the flow ratio calibrations have now moved into a dialog as shown below, instead of being separate menu items. Can you add the new option into this dialog as well?
<img width="267" height="379" alt="Screenshot 2026-05-08 at 12 04 14" src="https://github.com/user-attachments/assets/445e0c77-654c-443c-9fe8-1ad6f516f01c" />
This PR currently adds the menu item for Windows and Linux only, not for macOS. But since we’ll be moving the option to the dialog, this won’t be an issue once the change is made.

Some other suggestions — it would be better to have "Print Scale" and "Print Ruler" enabled by default. Otherwise it will be very difficult for new users. I really like your scale and text design BTW 👍 
<img width="404" height="582" alt="Screenshot 2026-05-08 at 12 09 43" src="https://github.com/user-attachments/assets/44e0bd34-75ba-4605-bbc3-fdae8e8f2298" />

pi-squared-studio commented:
> @pi-squared-studio Have you noticed that auto slice does not work? Moreover, if you select the interlaced option, it remains active even if you change the model or start a new project. 

Ок, I see. Yes, the problem remains because the calibration settings are not reset when the model is simply removed from the plate. If we create a new project, the calibration settings are not transferred. I will try to work around this issue, which is not user-friendly.

pi-squared-studio commented:
@SoftFever Ok. I just need to install the latest changes. I may have been working on an older version of Orca.

pi-squared-studio commented:
> One small change needed: the flow ratio calibrations have now moved into a dialog as shown below, instead of being separate menu items. Can you add the new option into this dialog as well?

I looked at this UI. The Practical Flow Test has a slightly different structure and logic than this set of classic tests.
Should I make a submenu for all flow tests, or should I combine them in a single dialog box?
<img width="416" height="247" alt="image" src="https://github.com/user-attachments/assets/cbd2e1f2-1989-4e80-988a-3acd7e9fb1a6" />

SoftFever commented:
> > One small change needed: the flow ratio calibrations have now moved into a dialog as shown below, instead of being separate menu items. Can you add the new option into this dialog as well?
> 
> I looked at this UI. The Practical Flow Test has a slightly different structure and logic than this set of classic tests. Should I make a submenu for all flow tests, or should I combine them in a single dialog box? <img alt="image" width="416" height="247" src="https://private-user-images.githubusercontent.com/189209038/589414594-cbd2e1f2-1989-4e80-988a-3acd7e9fb1a6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzgyMjY2ODIsIm5iZiI6MTc3ODIyNjM4MiwicGF0aCI6Ii8xODkyMDkwMzgvNTg5NDE0NTk0LWNiZDJlMWYyLTE5ODktNGU4MC05ODhhLTNhY2Q3ZTlmYjFhNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDUwOFQwNzQ2MjJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kY2U1NWY4NzEyMjliYjZmYjJmZjViYzYzZjVmZDljMDRjYzkwZWM1ZTY1NmU2NTZjZDc2YWM1NmIxNmZlNzE5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.uR4E1IR9tKdZtJMlm-V2FjIlWFHDV2iZb6x3nwf-pZM">

I suggest we add the practical flow ratio calibration as a new option in the dialog below.  The menu has removed and we shouldn't re-add them.
Here is a suggested workflow:
When the this calib method is selected, the top surface pattern can be set to the pattern you specified and then disabled so the user can't change it (since for this test, other patterns won't work).  
Once the user clicks "OK," it invokes your dialog for additional settings.
<img width="534" height="758" alt="image" src="https://github.com/user-attachments/assets/5be471d4-b709-40e3-bfff-af59ed32bce6" />

pi-squared-studio commented:
@SoftFever Okay, then I'll make the implementation, but then, if anything, with yours own preferences. However, the implementation of calling a new dialog box from another one is somewhat puzzling.
Yes, this test does not use a top surface fill template. The accumulation of excess material occurs within the material itself. It is similar to creating an object with 100% solid infill.
I will also make a number of additions for user convenience.

pi-squared-studio commented:
Here is a package of essential additions:
1. The "Print Rule" and "Print Scale" buttons are now enabled by default
2. The Autoslice button has appeared
<img width="432" height="594" alt="image" src="https://github.com/user-attachments/assets/1d5777e5-bccb-4052-a3c0-6a10c4865548" />

3. The scale "Rule" has been modified to display the current preset filament flow rate in the center of the range.
<img width="865" height="402" alt="image" src="https://github.com/user-attachments/assets/3075c6b8-5912-4d9f-a107-45346df01434" />

4. The settings that are passed to the print module have been modified. Previously, they applied to everything on the Plate, but now they only apply to the calibration model. Now, when you clean the bed and place another model, the calibration settings do not work. (I haven't checked, but it would be advisable to further investigate the impact of other tests, such as PA, on this conditions).

Yes, I couldn't manage to open a dialog from another dialog box, so I just added a line to the calibration menu. If someone more knowledgeable can make this menu more user-friendly, that would be great! However, the method of generating this test is different from other classic tests, so it seems questionable to combine this test with others in a single dialogue.
<img width="228" height="253" alt="image" src="https://github.com/user-attachments/assets/3e2bb1b6-d97d-4c40-aa6e-5a8d7e554ebd" />

Another unresolved issue. The printer and filament settings, particularly the z-Hop, are not passed to the object settings and will be shared across the entire Plate.

Later, I'll review the code again for any garbage lines that may have been introduced during multiple edits.

pi-squared-studio committed:
Big Updates

pi-squared-studio committed:
Final

pi-squared-studio commented:
<img width="986" height="509" alt="image" src="https://github.com/user-attachments/assets/7ea704e6-052d-48f7-ad65-4963490482c9" />

Another addition: if an asymmetric flow change range is entered, such as 0.95 and 1.10, and the main flow value is different from 1.0, the value will be displayed at a specific distance on the bottom scale. This is very useful for checking the calibration of the current flow.

@SoftFever, @RF47, @ianalexis  I looked at the garbage code, and I didn't see anything extra. If you like this PR, you can take it into use.

SoftFever committed:
Merge branch 'main' into Golden-Ratio-Flow-Calibration-Test

SoftFever commented on src/libslic3r/Fill/FillBase.cpp:
```diff
@@ -180,13 +191,82 @@ void Fill::fill_surface_extrusion(const Surface* surface, const FillParams& para
-        }
 
+        // Calibration section (post-extrusion) with sended parameters
+        if (this->calib_params != nullptr) {
```
this will effectively disalbe gap fill feature becuase with the new change `f->calib_params = &model_object()->calib_params;` will never be nullptr

pi-squared-studio replied:
This should be disabled by default, as calib_params is explicitly set by the test developer.
how it is implemented in src/slic3r/GUI/Plater.cpp

For this test:
<img width="775" height="175" alt="image" src="https://github.com/user-attachments/assets/0fad814b-616a-4095-a11c-1fd7e435de48" />

For classical flowrate tests
<img width="775" height="175" alt="image" src="https://github.com/user-attachments/assets/fa7b0c59-c968-4398-af87-a55f4f2e5fc2" />

For other tests, where there is no pre- and post-processing - calib_params == nullptr

SoftFever replied:
Can you double check?
this->calib_params will always be non nullptr

SoftFever commented on src/libslic3r/PrintConfig.cpp:
```diff
@@ -4118,11 +4118,6 @@ void PrintConfigDef::init_fff_params()
     def->mode     = comAdvanced;
     def->set_default_value(new ConfigOptionInt(2));
```
Removing option with handle legacy will case issues loading old .3mf file.
Is there any reason to remove it anyway?

SoftFever commented on src/libslic3r/Fill/FillBase.cpp:
```diff
@@ -162,12 +163,22 @@ void Fill::fill_surface_extrusion(const Surface* surface, const FillParams& para
+            case CalibMode::Calib_Flow_Rate:
+                if (params.extrusion_role == erTopSolidInfill) {
+                    eec->no_sort = true;
+                    break;
```
the break should move out of the if {} block

SoftFever commented on src/libslic3r/Fill/FillBase.cpp:
```diff
@@ -162,12 +163,22 @@ void Fill::fill_surface_extrusion(const Surface* surface, const FillParams& para
+                    eec->no_sort = true;
+                    break;
+                }
+            case CalibMode::Calib_Practical_Flowratio: 
```
add break for this case too

SoftFever reviewed (changes requested):
hi please take a look at the comments

pi-squared-studio committed:
Update FillBase.cpp

MxBrnr commented:
I think the test should be renamed. Calling it the "Practical" test can be rather misleading and is too likely to bias users into thinking its results should fully supercede those of the other tests.

pi-squared-studio commented:
> I think the test should be renamed. Calling it the "Practical" test can be rather misleading and is too likely to bias users into thinking its results should fully supercede those of the other tests.

I don't mind. By the way, I've already renamed it once... The problem with the name is that I'm not a native speaker, and I can't take into account cultural and technical aspects in the names.

Personally, I don't care what this test is called. Let the community decide what's best.

Felix14-v2 commented:
Maybe something like "single-sample flow ratio test", "flow ratio gradient test" or  "flow ratio incremental test"?

But I'm not a native speaker too :)

boromyr commented:
Instead of practical, we could call it quick or rapid, since it is a quick test. Or even minimal.

SoftFever commented:
How about call it "Continuous Flow Ratio Test".
Nevertheless, we should move it into the dialog instead of in the menu item.

pi-squared-studio commented:
> How about call it "Continuous Flow Ratio Test". Nevertheless, we should move it into the dialog instead of in the menu item.

You can choose this name as the project organizer. For me, this will be like a marker for marking technologies, for further explanation in the description.

Regarding the menu, it will be difficult for me to implement it because I do not know the wxWidgets structure very well. It turns out that you need to open dialog box 2 from dialog box 1, and then use the latter to control the correct closing of the former and start the calibration procedure. At this point, it seems that I will have problems with this organization of actions. Can someone who is more experienced help me create this interface?

On the other hand, is there an idea to collect all calibration tests with similar actions under one submenu? In the future, it will be useful for users if the number of such tests increases.

<img width="254" height="283" alt="image" src="https://github.com/user-attachments/assets/00772766-7735-4eff-9b75-8e308679410d" />

pi-squared-studio commented:
@SoftFever I think I've answered all the requests, and everything should be fine in the code. Tell me, do I need to modify the code regarding the name? At the moment, I can't do it quickly because Git is set to another unsaved PR that I want to finish. 
If you rename the project yourself, that would be great.

As for the interface, can we ask @ianalexis, @RF47, @yw4z, or someone else with experience to design the interface into the new PR. As I mentioned earlier, I'm hesitant to design the UI because it requires a higher level of expertise.

no1geek commented:
anything new on it? I am waiting for this feature to be merged.
Where can I download a build of it?

pi-squared-studio committed:
Merge branch 'main' into Golden-Ratio-Flow-Calibration-Test

pi-squared-studio committed:
Fix

pi-squared-studio commented:
@ianalexis I want to complete all my open projects, and I would like to get them working. I wanted to ask about this PR? Should I rename it test to Advanced or give it a different unique name?

ianalexis commented:
What about something like "Progressive Bar" or something like that?

pi-squared-studio commented:
> What about something like "Progressive Bar" or something like that?

I want to choose the simplest word to distinguish it from the Standard test. Yes, I'll also modify the menu to provide easy access to all the tests. It won't be possible to combine them under a single dialog box because the base of these tests is different.

pi-squared-studio committed:
Restore Functionality + Check of CoPilot msgs

pi-squared-studio commented:
I tested it for functionality. There was an error related to changing configuration rules in Orca v2.5.0, so it crashed.

I also noticed that the test only works with the first extruder. I think I'll fix the ability to select it before the release if multi-head systems are used.

pi-squared-studio commented:
> What about something like "Progressive Bar" or something like that?

Ok, for now I'll just make it Progressive - Progressive Flow Ratio Calibration Test  (PFRCT)

pi-squared-studio commented:
<img width="553" height="224" alt="image" src="https://github.com/user-attachments/assets/efcbba94-5e9a-4146-a706-e9d34ea403d5" />

In the last edition, I will leave this menu. I decided to group them by common feature. I think it's better this way. It will be especially useful if later someone add submenu by commands, for ex, to run calibration tests using G-commands from Orca.

Can also sort these tests by basical-advanced, but that's a topic for another PR.

pi-squared-studio committed:
Functionality

pi-squared-studio committed:
Merge branch 'main' into Golden-Ratio-Flow-Calibration-Test

pi-squared-studio committed:
Look up

pi-squared-studio committed:
Merge branch 'Golden-Ratio-Flow-Calibration-Test' of https://github.com/pi-squared-studio/OrcaSlicer into Progressive-Flow-Ratio-Calibration-Test

pyramor commented:
> > What about something like "Progressive Bar" or something like that?
> 
> Ok, for now I'll just make it Progressive - Progressive Flow Ratio Calibration Test (PFRCT)

What about just "Simple" Flow Ratio Calibration Test?

ianalexis commented:
> <img alt="image" width="553" height="224" src="https://private-user-images.githubusercontent.com/189209038/628578483-efcbba94-5e9a-4146-a706-e9d34ea403d5.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODUzNDIyNDIsIm5iZiI6MTc4NTM0MTk0MiwicGF0aCI6Ii8xODkyMDkwMzgvNjI4NTc4NDgzLWVmY2JiYTk0LTVlOWEtNDE0Ni1hNzA2LWU5ZDM0ZWE0MDNkNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNzI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDcyOVQxNjE5MDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iYzBiOGE3Yjg4YjE2MDhjZGM1Zjg2Y2NiNTdlZjZlY2ZiMjNhYjY5ZWM1ZTY1MjAwZTc0ZmQzMDUwNzRhYWM2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.3L4sMW4t43hhYyxBHxOoDAMkYr38TDw46Q6bHjbpSMs">
> 
> In the last edition, I will leave this menu. I decided to group them by common feature. I think it's better this way. It will be especially useful if later someone add submenu by commands, for ex, to run calibration tests using G-commands from Orca.
> 
> Can also sort these tests by basical-advanced, but that's a topic for another PR.

Keep max flowrate as its own section as it currently is. Is a different test concept and it has to be done before PA following this order https://www.orcaslicer.com/wiki/guides/calibration_guide.html



> What about just "Simple" Flow Ratio Calibration Test?

the thing about "simple" is that is "simple" for who? is better to describe it the best we can.

That's why i said "Bar" and maybe the other has to be "tiles"(?)

pi-squared-studio commented:
> What about just "Simple" Flow Ratio Calibration Test?

Personally, I don't care what it's called. What matters to me is that the product works.

Regarding "Simple" - I think it's not all users to simple understand and interpret the results, because this test has its own unique features. It may be faster and more compact, but not simple than classical...

pi-squared-studio commented:
> Keep max flowrate as its own section as it currently

OK, I'm doing it.

> it has to be done before PA following

Following this logic, Flow Rate is the second test after Temperature, as it is the most basic test on which all other tests depend. But I won't change it in this PR

pi-squared-studio committed:
Return the menus order

ianalexis commented:
> Following this logic, Flow Rate is the second test after Temperature, as it is the most basic test on which all other tests depend. But I won't change it in this PR

The logic here is that for flow rate you need a good PA and for good PA you need a good max volumetric flow limit.

That's why this "recommended" order.

pi-squared-studio committed:
Update FillBase.cpp

pi-squared-studio commented:
> > Following this logic, Flow Rate is the second test after Temperature, as it is the most basic test on which all other tests depend. But I won't change it in this PR
> 
> The logic here is that for flow rate you need a good PA and for good PA you need a good max volumetric flow limit.
> 
> That's why this "recommended" order.

<img width="458" height="258" alt="image" src="https://github.com/user-attachments/assets/c98493cc-c4ba-4847-b01e-55fa77c0edc4" />

I'm done with this project. It can be merged. I'll leave the choice of the name to you as you like. 
The code can be easily corrected using auto-replace with the search term "Progressive".

ianalexis commented:
Nice, will test it this keep i promise hahahah

pi-squared-studio committed:
Merge branch 'main' into Golden-Ratio-Flow-Calibration-Test
